mat64/io.go

```diff
@@ -82,28 +68,73 @@ func (m *Dense) UnmarshalBinary(data []byte) error {
 	}
 
 	buf := bytes.NewReader(data)
-	var rows int64
-	err := binary.Read(buf, binary.LittleEndian, &rows)
+	n, err := m.UnmarshalBinaryFrom(buf)
 	if err != nil {
 		return err
 	}
+	if n != len(data) {
+		return errBadBuffer
+	}
+	if len(data) != len(m.mat.Data)*sizeFloat64+2*sizeInt64 {
+		return errBadBuffer
+	}
+
+	return nil
+}
+
+func (m Dense) MarshalBinaryTo(w io.Writer) (int, error) {
+	var n int
+	err := binary.Write(w, binary.LittleEndian, int64(m.mat.Rows))
+	if err != nil {
+		return n, err
+	}
+	n += sizeInt64
+	err = binary.Write(w, binary.LittleEndian, int64(m.mat.Cols))
+	if err != nil {
+		return n, err
+	}
+	n += sizeInt64
+	r, c := m.Dims()
+	for i := 0; i < r; i++ {
+		for j := 0; j < c; j++ {
+			err = binary.Write(w, binary.LittleEndian, m.at(i, j))
```

Comment:
Actually, for the read operation don't use set, just iterate across the Data slice since we know the rows are contiguous: `m.mat.Data[i] = math.Float64frombits(binary.LittleEndian.Uint64(buf[:]))`